core/src/main/java/com/linecorp/armeria/internal/metric/PrometheusMetricRequestDecorator.java

```diff
@@ -0,0 +1,370 @@
+     * @return A service decorator function
+     */
+    public static <T extends MetricLabel<T>, I extends Request, O extends Response>
+    PrometheusMetricRequestDecorator<T, I, O> decorateService(
+            CollectorRegistry collectorRegistry,
+            Iterable<T> metricLabels,
+            Function<RequestLog, Map<T, String>> labelingFunction) {
+        return new PrometheusMetricRequestDecorator<>(collectorRegistry,
+                                                      metricLabels,
+                                                      labelingFunction,
+                                                      "armeria_server");
+    }
+
+    /**
+     * Returns a new {@link Client} decorator that tracks request stats using the Prometheus metrics  library.
+     *
+     * @param <I> Request type
+     * @param <O> Response type
+     * @param collectorRegistry Prometheus registry
+     * @param metricLabels Labels names for metrics
+     * @param labelingFunction map of labels to values
+     * @return A client decorator function
+     */
+    public static <T extends MetricLabel<T>, I extends Request, O extends Response>
+    PrometheusMetricRequestDecorator<T, I, O> decorateClient(
+            CollectorRegistry collectorRegistry,
+            T[] metricLabels,
+            Function<RequestLog, Map<T, String>> labelingFunction) {
+        return new PrometheusMetricRequestDecorator<>(collectorRegistry,
+                                                      metricLabels,
+                                                      labelingFunction,
+                                                      "armeria_client");
+    }
+
+    /**
+     * Returns a new {@link Client} decorator that tracks request stats using the Prometheus metrics  library.
+     *
+     * @param <I> Request type
+     * @param <O> Response type
+     * @param collectorRegistry Prometheus registry
+     * @param metricLabels Labels names for metrics
+     * @param labelingFunction map of labels to values
+     * @return A client decorator function
+     */
+    public static <T extends MetricLabel<T>, I extends Request, O extends Response>
+    PrometheusMetricRequestDecorator<T, I, O> decorateClient(
+            CollectorRegistry collectorRegistry,
+            Iterable<T> metricLabels,
+            Function<RequestLog, Map<T, String>> labelingFunction) {
+        return new PrometheusMetricRequestDecorator<>(collectorRegistry,
+                                                      metricLabels,
+                                                      labelingFunction,
+                                                      "armeria_client");
+    }
+
+    private final Function<RequestLog, Map<T, String>> labelingFunction;
+    private final MetricAdapter metricAdapter;
+
+    private PrometheusMetricRequestDecorator(
+            CollectorRegistry collectorRegistry,
+            Object[] metricLabels,
```

Comment:
I'm not sure, but I think because of the bounds on the type parameter, this ends up being a safe cast. If it isn't, it should still at least be safe using `MetricLabel[]` instead of `Object[]`.

```java
(T[]) Iterables.toArray(requireNonNull(metricLabels, "metricLabels"), MetricLabel.class)
```
I applied this patch to your branch and the integration test still passes - https://gist.github.com/anuraaga/7c37e012fcd260c4da2bc0963ab79315